examples/common/tracing/TraceDecoder.cpp

```diff
@@ -0,0 +1,414 @@
+
+    builder.Add(" / Exchange = ");
+    builder.Add(static_cast<uint16_t>(json[kExchangeIdKey].asLargestUInt()));
+    builder.Add("]");
+
+    ChipLogProgress(DataManagement, "%s", builder.c_str());
+
+    json.removeMember(kSourceKey);
+    json.removeMember(kDestinationKey);
+    json.removeMember(kSessionIdKey);
+    json.removeMember(kExchangeIdKey);
+    json.removeMember(kMessageCounterKey);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TraceDecoder::MaybeLogAndConsumePayload(Json::Value & json, bool isResponse)
+{
+    auto size = static_cast<uint16_t>(json[kPayloadSizeKey].asLargestUInt());
+    if (size)
+    {
+        bool shouldDecode = !isResponse || mOptions.mEnableProtocolInteractionModelResponse;
+        auto payload      = json[kPayloadDataKey].asString();
+        auto protocolId   = json[kProtocolIdKey].asLargestUInt();
+        auto protocolCode = json[kProtocolCodeKey].asLargestUInt();
+        ReturnErrorOnFailure(LogAsProtocolMessage(protocolId, protocolCode, payload.c_str(), payload.size(), shouldDecode));
+        ChipLogDetail(DataManagement, "");
+    }
+
+    json.removeMember(kPayloadDataKey);
+    json.removeMember(kPayloadSizeKey);
+    json.removeMember(kProtocolIdKey);
+    json.removeMember(kProtocolCodeKey);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TraceDecoder::MaybeLogAndConsumeSecurityFlags(Json::Value & json)
+{
+    VerifyOrReturnError(json.isMember(kSecurityFlagsKey), CHIP_NO_ERROR);
+
+    auto flags = static_cast<uint8_t>(json[kSecurityFlagsKey].asLargestUInt());
+    if (flags)
+    {
+        ChipLogDetail(DataManagement, "    Security (0x%02x) =", flags);
+        ChipLogDetail(DataManagement, "    {");
+
+        if (flags & to_underlying(chip::Header::SecFlagValues::kPrivacyFlag))
+        {
+            ChipLogDetail(DataManagement, "        Privacy = true");
+        }
+
+        if (flags & to_underlying(chip::Header::SecFlagValues::kControlMsgFlag))
+        {
+            ChipLogDetail(DataManagement, "        ControlMsg = true");
+        }
+
+        if (flags & to_underlying(chip::Header::SecFlagValues::kMsgExtensionFlag))
+        {
+            ChipLogDetail(DataManagement, "        MsgExtension = true");
+        }
+
+        ChipLogDetail(DataManagement, "    }");
+    }
+
+    json.removeMember(kSecurityFlagsKey);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TraceDecoder::MaybeLogAndConsumeMessageFlags(Json::Value & json)
+{
+    VerifyOrReturnError(json.isMember(kMessageFlagsKey), CHIP_NO_ERROR);
+
+    auto flags = static_cast<uint8_t>(json[kMessageFlagsKey].asLargestUInt());
+    if (flags)
+    {
+        ChipLogDetail(DataManagement, "    Message (0x%02x) =", flags);
+        ChipLogDetail(DataManagement, "    {");
+
+        if (flags & to_underlying(chip::Header::MsgFlagValues::kSourceNodeIdPresent))
+        {
+            auto id = json[kSourceNodeIdKey].asLargestUInt();
+            ChipLogDetail(DataManagement, "        SourceNodeId = 0x%16llx", id);
+        }
+
```

Comment:
Why are we converting to some bespoke format, rather than direct JSON that can be read by the test harness Python code?

Why are we writing C++ code to parse JSON, rather than writing Python code for the TH ?

Reply:
> Why are we converting to some bespoke format, rather than direct JSON that can be read by the test harness Python code?
> Why are we writing C++ code to parse JSON, rather than writing Python code for the TH ?

So we are speaking of different things here. You mention `Python code for the TH`,  but this PR is not about TH.

That said, and even if it was, most of the TH things used by the CSG that I know off are C++ tools (chip-tool / examples apps / YAML tests suites). I'm not sure about what you are referring too when you say `Python code for the TH ` ?

So this PR is about:
 * A command line tool to read the trace (so bespoke format)
 * Logs that integrates natively into the current debugging tool that is used by developers and the CSG (chip-tool / all examples apps under Linux/Darwin) (so bespoke format and C++)